Inflection 85: Add support for Malayalam

deonajulary06 opened the pull request:
The project expands Unicode Inflection to support Malayalam, allowing for grammatical inflections for nouns, pronouns, verbs, adjectives, and any other relevant words. Using data from Wikidata and custom inflection rules implemented in C++ and C, the project ensures Malayalam is correctly formatted across number, gender, case, tense, mood, and other aspects. This enhances language tools, making communication in Malayalam more accurate and accessible.

Includes Fixes #85

deonajulary06 committed:
# This is a combination of 25 commits.
# This is the 1st commit message:

Add Malayalam dictionary support

# This is the commit message #2:

Add Malayalam language support in LocaleUtils.hpp

# This is the commit message #3:

Add Malayalam locale support to LocaleUtils

# This is the commit message #4:

Add Malayalam language to the tests

# This is the commit message #5:

Add Malayalam locale group ml_IN

# This is the commit message #6:

ADD: Malayalam tokenizer configuration file

# This is the commit message #7:

Inflection-85: Add Git LFS config for Malayalam dictionary and XML files

# This is the commit message #8:

Add Malayalam inflection and pronoun tests

# This is the commit message #9:

Updated copyright line

# This is the commit message #10:

Updated copyright message

# This is the commit message #11:

Updated copright message

# This is the commit message #12:

Updated language grammar to include Malayalam

# This is the commit message #13:

Added pronouns for Malayalam

# This is the commit message #14:

Add ll GrammarSynthesizer files

# This is the commit message #15:

Add Malayalam grammar synthesizer

# This is the commit message #16:

Add Malayalam-specific CommonConceptFactory with lists and quantities

# This is the commit message #17:

Update document on how to add a new language, fixed errors

# This is the commit message #18:

Updated grammar.xml for Malayalam

# This is the commit message #19:

Update pronoun_ml.csv

# This is the commit message #20:

Updated all grammar synthesizer component for Malayalam

# This is the commit message #21:

Update Common Concept Factory files

# This is the commit message #22:

Updated tests for Malayalam

# This is the commit message #23:

Fix Malayalam grammar synthesis and remove count lookup function

# This is the commit message #24:

Updated Grammeme Constants files to include sociative case

# This is the commit message #25:

Temporary fix for GitHub

FrankYFTang committed:
Inflection-120 Update ICU version in versions.mk to 77.1 (#119)

* Update versions.mk

* Update cmake-multi-platform.yml

* Update cmake-multi-platform.yml

* Update cmake-multi-platform.yml

FrankYFTang committed:
Add Factory for MF2 ( issues/117 ) (#114)

* Add Factory for MF2

Also add unit test and data driven unit test
Test data based on
inflection/test/resources/inflection/dialog/inflection/*.xml

Test formatter while the <result> is not empty
Test Selector while the <result> is empty, and there are one attribute
which is not "exists".

Uppercase the factory function

Change copyright, remove iostream, add comments

Fix build issue if the icu is < 77

Remove Data Driven test

Fix Locale building

Remove version Check

* Reformat

FrankYFTang committed:
Add checkleak ci to use valgrind. Fix issues/118 (#121)

grhoten reviewed (commented):
This looks like good progress. I've added some comments to help guide you to fix some test failures.

FrankYFTang committed:
Fix memory leak in MF2Factory.cpp (issue/131) (#132)

* Fix memory leak in MF2Factory.cpp (issue/131)

Need to delete the return of toSpeakableString() and getFeatureValue()
return from InflectableStringConcept.

Fix ICU cache

* Use std::unique_ptr instead of raw pointer and delete

* Fix per george feedback

FrankYFTang committed:
Issue 134, use std::unique_ptr in SemanticFeatureModel (#137)

FrankYFTang committed:
Issue 134, use std::unique_ptr in (De|It)CommonConceptFactory.cpp (#136)

FrankYFTang committed:
Issue 134, use std::unique_ptr in RuCommonConceptFactory.cpp (#135)

FrankYFTang committed:
Inflection 134 std::unique_ptr for  Chunk.cpp (#145)

FrankYFTang committed:
Inflection 134 std::unique_ptr for  TokenExtractor (#144)

* Inflection 134 std::unique_ptr for  TokenExtractor

* remove ()

* remove ()

FrankYFTang committed:
Inflection 140: Change the build process to also build on merging (#139)

* Update cmake-multi-platform.yml

* Update ubuntu-memory-check.yml

* Update ubuntu-memory-check.yml

* Update cmake-multi-platform.yml

FrankYFTang committed:
Inflection 134 std::unique_ptr for DictionaryMetaData.cpp (#143)

* Inflection 134 std::unique_ptr for DictionaryMetaData.cpp

* fix

FrankYFTang committed:
Inflection 134 std::unique_ptr for DictionaryKeyIterator.cpp (#142)

* Inflection 134 std::unique_ptr for DictionaryKeyIterator.cpp

* add npc

FrankYFTang committed:
Inflection 134 std::unique_ptr for KoGrammarSynthesizer_ParticleResol… (#141)

* Inflection 134 std::unique_ptr for KoGrammarSynthesizer_ParticleResolver.cpp

* add npc

FrankYFTang committed:
Cache LFS content (fix issue/127) (#128)

* Cache LFS content (fix issue/127)

and fix icu cache by adding save action

* only perform lfs action if missed the cache

* Add ls to debug the content restored

* Debug

* Correct path

* fix

* Remove dupicate lines

FrankYFTang committed:
WiP Add doxygen (#146)

* Try to add doxygen

* Update doxygen

* Update file

* Fix Doxygen file

* Add dos gen

* Skip mac for doc gen

* add doxgey

* Use DenverCoder1/doxygen-github-pages-action

* Tweak the configue path

* Try to fix dependency

* Add different yml for deploy

* Remove extra lines

* Add push

* add debug

* Change path

* Fix

* Change way to install doxygen

* Fix

* Add run to debug

* push

* Remove double $

* Remove

FrankYFTang committed:
Change namespace in message2 (issue/147) (#148)

* Change namespace in message2

* Remove namespace in hpp file

FrankYFTang committed:
Fix doxygen configuraiton (#149)

Here is why we fix the doxygen in this way.

I discussed with Gemini about how to fix this issue by asking

why doxygen cannot recognize class like class INFLECTION_CLASS_API inflection::dialog::NumberConcept ? how to fix it?

Gemini answered:

Doxygen's primary job is to parse C++, C, Java, Python, and other code to generate documentation. When Doxygen fails to recognize a class like class INFLECTION_CLASS_API inflection::dialog::NumberConcept, it's almost always due to the preprocessor macro INFLECTION_CLASS_API.

Here's why this happens and how to fix it:

Why Doxygen Cannot Recognize It

Preprocessor Macro Obscuring the Class Definition: Doxygen has a built-in preprocessor. If INFLECTION_CLASS_API is a macro that expands to something like __declspec(dllexport) (common for Windows DLLs) or __attribute__((visibility("default"))) (common for GCC/Clang on Linux/macOS), Doxygen's parser, by default, might not correctly interpret the resulting code. It sees __declspec(dllexport) class ... instead of class ....

ENABLE_PREPROCESSING = NO: If ENABLE_PREPROCESSING is set to NO in your Doxyfile, Doxygen won't process any macros, and it will simply see INFLECTION_CLASS_API as part of the class name, leading to a parsing error.

Macro Expansion Issues: Even with preprocessing enabled, Doxygen's default macro expansion behavior might not be sufficient. It might not expand INFLECTION_CLASS_API in a way that reveals the underlying class keyword to its parser.

How to Fix It

The solution involves configuring Doxygen's preprocessor to correctly handle the INFLECTION_CLASS_API macro.

Enable Preprocessing:
Make sure ENABLE_PREPROCESSING is set to YES in your Doxyfile.

Code snippet

ENABLE_PREPROCESSING = YES
Define the Macro in Doxygen:
You need to tell Doxygen what INFLECTION_CLASS_API means so it can effectively ignore or replace it during parsing. This is done using the PREDEFINED tag in your Doxyfile.

If INFLECTION_CLASS_API is empty (or should be ignored by Doxygen):
This is often the case for API macros where they expand to nothing for documentation builds.

Code snippet

PREDEFINED = INFLECTION_CLASS_API=
This tells Doxygen to replace INFLECTION_CLASS_API with nothing.

If INFLECTION_CLASS_API expands to a __declspec or __attribute__:
You need to remove these compiler-specific attributes.
For __declspec(dllexport) or __declspec(dllimport):

Code snippet

PREDEFINED = __declspec(x)=
This tells Doxygen to replace __declspec(anything) with nothing.

For __attribute__((visibility("default"))) or similar:

Code snippet

PREDEFINED = __attribute__(x)=
This tells Doxygen to replace __attribute__(anything) with nothing.

If INFLECTION_CLASS_API is a more complex macro:
If it's something like #define INFLECTION_CLASS_API SOME_OTHER_MACRO class, you might need to combine PREDEFINED with MACRO_EXPANSION and EXPAND_ONLY_PREDEF.

For example, if INFLECTION_CLASS_API itself expands to a simple empty string or a known attribute, the above methods should work. If it's more complex, you might need to use MACRO_EXPANSION = YES and EXPAND_ONLY_PREDEF = YES, then explicitly define INFLECTION_CLASS_API in PREDEFINED to be empty or a simple class keyword if it's truly hiding class. However, it's generally better to make the macro itself expand to an empty string for Doxygen.

Example PREDEFINED entry (most common scenario):

Code snippet

PREDEFINED = INFLECTION_CLASS_API= \
             __declspec(x)= \
             __attribute__(x)=
(Note the \ for multi-line definitions).

Run Doxygen in Debug Mode (Preprocessor Output):
If the above doesn't work, you can see exactly what Doxygen's preprocessor is doing by running Doxygen with the -d Preprocessor flag:

Bash

doxygen -d Preprocessor your_Doxyfile
This will print the preprocessed output to the console, allowing you to see how INFLECTION_CLASS_API is being handled and debug your PREDEFINED settings.

Steps to Implement the Fix:

Open your Doxyfile: This is the configuration file for your Doxygen project.

Locate ENABLE_PREPROCESSING: Ensure it's set to YES. If not, change it.

Locate PREDEFINED: Add or modify this tag.

Find out what INFLECTION_CLASS_API expands to in your actual build system (e.g., by looking at compiler preprocessor output or just checking the macro definition in your code).

Based on that, add the appropriate definition to PREDEFINED to effectively remove or normalize INFLECTION_CLASS_API for Doxygen.

Re-run Doxygen: After modifying the Doxyfile, run Doxygen again to generate the documentation.

By correctly configuring Doxygen's preprocessor to handle INFLECTION_CLASS_API, you should be able to get it to recognize your inflection::dialog::NumberConcept class.

FrankYFTang committed:
Correct path to README.md and include that into the API doc. (#150)

FrankYFTang committed:
Fix some warning in doxygen (issue 147) (#151)

FrankYFTang committed:
Update README.md (#153)

FrankYFTang committed:
Add comment for super class documentation Fix 147  (#152)

* Add comment for super class documentation Fix 147

* Fix by using @cond/@endcond

* change // to ///

FrankYFTang committed:
Fix doxygen warning for DisplayValue (#154)

add documentation for operator ==, != and ()

FrankYFTang committed:
fix mistake in PR 152 (issue 147) (#156)

correct the name of the parameter o -> displayValue

FrankYFTang committed:
Update DisplayValue.hpp (#157)

FrankYFTang committed:
fix doxygen warning in SemanticFeatureModel_DisplayData.hpp (issue 147) (#158)

* fix doxygen warning in SemanticFeatureModel_DisplayData.hpp

Add document for operator ==, != and ()

* Update SemanticFeatureModel_DisplayData.hpp

FrankYFTang committed:
Fix the doxygen warnings (issue 147) (#159)

* Update InflectableStringConcept.hpp

* Update SemanticConceptList.hpp

* Update SemanticFeature.hpp

* Update SemanticFeature.hpp

* Update SemanticFeatureConceptBase.hpp

* Update SemanticFeatureModel.hpp

* Update SemanticValue.hpp

* Update SpeakableString.hpp

* Update XMLParseException.hpp

* Update LanguageGrammarFeatures_Feature.hpp

* Update LanguageGrammarFeatures_GrammarCategory.hpp

* Update LanguageGrammarFeatures_GrammarFeatures.hpp

* Update LanguageGrammarFeatures_GrammarFeatures.hpp

* Update LanguageGrammarFeatures_GrammarFeatures.hpp

* Update inflection/src/inflection/dialog/InflectableStringConcept.hpp

Co-authored-by: George Rhoten <grhoten@users.noreply.github.com>

* Update inflection/src/inflection/dialog/SemanticFeature.hpp

Co-authored-by: George Rhoten <grhoten@users.noreply.github.com>

* Update inflection/src/inflection/dialog/SemanticFeature.hpp

Co-authored-by: George Rhoten <grhoten@users.noreply.github.com>

* Update inflection/src/inflection/dialog/SemanticFeatureConceptBase.hpp

Co-authored-by: George Rhoten <grhoten@users.noreply.github.com>

* Update inflection/src/inflection/dialog/SemanticFeatureConceptBase.hpp

Co-authored-by: George Rhoten <grhoten@users.noreply.github.com>

* Update inflection/src/inflection/dialog/SemanticFeatureConceptBase.hpp

Co-authored-by: George Rhoten <grhoten@users.noreply.github.com>

* Update inflection/src/inflection/dialog/SemanticFeatureConceptBase.hpp

Co-authored-by: George Rhoten <grhoten@users.noreply.github.com>

* Update inflection/src/inflection/dialog/SemanticFeatureConceptBase.hpp

Co-authored-by: George Rhoten <grhoten@users.noreply.github.com>

* Update inflection/src/inflection/dialog/SemanticFeatureModel.hpp

Co-authored-by: George Rhoten <grhoten@users.noreply.github.com>

* Update inflection/src/inflection/dialog/SemanticFeatureModel.hpp

Co-authored-by: George Rhoten <grhoten@users.noreply.github.com>

* Update inflection/src/inflection/dialog/SemanticFeatureModel.hpp

Co-authored-by: George Rhoten <grhoten@users.noreply.github.com>

* Update inflection/src/inflection/dialog/SemanticFeatureModel.hpp

Co-authored-by: George Rhoten <grhoten@users.noreply.github.com>

* Update inflection/src/inflection/dialog/SemanticFeatureModel.hpp

Co-authored-by: George Rhoten <grhoten@users.noreply.github.com>

* Update inflection/src/inflection/dialog/SemanticFeatureModel.hpp

Co-authored-by: George Rhoten <grhoten@users.noreply.github.com>

* Update inflection/src/inflection/dialog/SemanticValue.hpp

Co-authored-by: George Rhoten <grhoten@users.noreply.github.com>

* Update inflection/src/inflection/dialog/SpeakableString.hpp

Co-authored-by: George Rhoten <grhoten@users.noreply.github.com>

* Update inflection/src/inflection/exception/XMLParseException.hpp

Co-authored-by: George Rhoten <grhoten@users.noreply.github.com>

* Update inflection/src/inflection/lang/features/LanguageGrammarFeatures_Feature.hpp

Co-authored-by: George Rhoten <grhoten@users.noreply.github.com>

* Update inflection/src/inflection/dialog/SemanticFeatureConceptBase.hpp

Co-authored-by: George Rhoten <grhoten@users.noreply.github.com>

* Update SemanticFeature.hpp

formatting

* Update inflection/src/inflection/dialog/SemanticFeatureConceptBase.hpp

Co-authored-by: George Rhoten <grhoten@users.noreply.github.com>

* Update SemanticFeatureConceptBase.hpp

formatting

* Update SemanticFeatureModel.hpp

* Update LanguageGrammarFeatures_Feature.hpp

* Update SemanticFeature.hpp

* Update LanguageGrammarFeatures_GrammarCategory.hpp

* Update LanguageGrammarFeatures_GrammarFeatures.hpp

* Update SemanticFeatureConceptBase.hpp

* Update SemanticFeature.hpp

* Update LanguageGrammarFeatures_GrammarFeatures.hpp

* Update LanguageGrammarFeatures_GrammarCategory.hpp

* Update LanguageGrammarFeatures_GrammarCategory.hpp

---------

Co-authored-by: George Rhoten <grhoten@users.noreply.github.com>

dockerh committed:
Create distribution packages for ubuntu (.deb release v0.0.1) (#129)

* Github workflow for Ubuntu Packaging

* Github workflow for Ubuntu Packaging.

* Github workflow for Ubuntu Packaging!

* Refactor packaging workflow based on the feedback

* Fix CPack versioning: derive major/minor/patch from INFLECTION_VERSION

* Fixing Version Issue

* Fixing the Cmake Version

* Fix: Cpack Version Handling

* Add CPack-based Ubuntu packaging and GitHub Actions release workflow

FrankYFTang committed:
issue 163 fix ULocale < and > operator order (#164)

dockerh committed:
Fix GitHub Release Trigger, ICU Cache Cleanup, and Artifact Upload  (#166)

* Github workflow for Ubuntu Packaging

* Github workflow for Ubuntu Packaging.

* Github workflow for Ubuntu Packaging!

* Refactor packaging workflow based on the feedback

* Fix CPack versioning: derive major/minor/patch from INFLECTION_VERSION

* Fixing Version Issue

* Fixing the Cmake Version

* Fix: Cpack Version Handling

* Add CPack-based Ubuntu packaging and GitHub Actions release workflow

* Fix GitHub release workflow, fix cache key and release upload

* Update workflow trigger to release on v* tags

* Fix release workflow: use stable ICU cache key

* Fix release workflow - Ubuntu

* Fix release workflow - Ubuntu Packaging

* Fix Ubuntu packaging workflow: ensure ICU env vars are exported for all build/test steps

FrankYFTang committed:
Add operator<=> and remove operator< operator and compareTo function (issue 155) (#160)

* Remove operator< and compareTo and add operator<=>

* Update ULocale.hpp

* Fix hash test. Ensure the comparator sort correctly.

FrankYFTang committed:
Inflection 134 Use std::unique_ptr to avoid delete (#162)

* Use std::unique_ptr to avoid delete

* Remove change to warm up code

* Make destructor public instead of adding friend

FrankYFTang committed:
Fail doxygen warning (#161)

* Turn on fail on doxygen warning

* Turn on runnong doxygen on PR

* Fix images error

* Update doxygen-gh-pages.yml

grhoten committed:
Inflection-63 Integrate ko Wikidata into Unicode Inflection Inflection-62 Integrate ar Wikidata into Unicode Inflection Inflection-61 Integrate he Wikidata into Unicode Inflection Inflection-60 Integrate hi Wikidata into Unicode Inflection Inflection-58 Integrate nb Wikidata into Unicode Inflection Inflection-56 Integrate nl Wikidata into Unicode Inflection Inflection-55 Integrate tr Wikidata into Unicode Inflection Inflection-54 Integrate ru Wikidata into Unicode Inflection Inflection-53 Integrate it Wikidata into Unicode Inflection Inflection-52 Integrate pt Wikidata into Unicode Inflection Inflection-51 Integrate fr Wikidata into Unicode Inflection Inflection-50 Integrate de Wikidata into Unicode Inflection (#167)

Inflection-63 Integrate ko Wikidata into Unicode Inflection
Inflection-62 Integrate ar Wikidata into Unicode Inflection
Inflection-61 Integrate he Wikidata into Unicode Inflection
Inflection-60 Integrate hi Wikidata into Unicode Inflection
Inflection-58 Integrate nb Wikidata into Unicode Inflection
Inflection-56 Integrate nl Wikidata into Unicode Inflection
Inflection-55 Integrate tr Wikidata into Unicode Inflection
Inflection-54 Integrate ru Wikidata into Unicode Inflection
Inflection-53 Integrate it Wikidata into Unicode Inflection
Inflection-52 Integrate pt Wikidata into Unicode Inflection
Inflection-51 Integrate fr Wikidata into Unicode Inflection
Inflection-50 Integrate de Wikidata into Unicode Inflection

FrankYFTang committed:
Fix incorrect cache key (#170)

* Update doxygen-gh-pages.yml

* Update cmake-multi-platform.yml

* Update ubuntu-memory-check.yml

* Update cmake-multi-platform.yml

* Update doxygen-gh-pages.yml

* Update ubuntu-memory-check.yml

* Update create-ubuntu-distribution-packaging.yml

* Update create-ubuntu-distribution-packaging.yml

fix comments

* Update cmake-multi-platform.yml

* Update doxygen-gh-pages.yml

grhoten commented on documents/how_to_add_new_language.md:
```diff
@@ -10,6 +10,9 @@ The following steps with help you identify files that need to be added or change
 NOTE: Take a look at [PR #40](https://github.com/unicode-org/inflection/pull/40) and [PR #111](https://github.com/unicode-org/inflection/pull/111) for example on how to add initial language support based on dictionary lookup only.
 In general, to bootstrap your progress look for grammatically similar language that's already supported, e.g. if you are adding Serbian look for existing Russian implementation.
 This will help you find most of the files you need to add/change and will speed up implementation of the rules and lexicons.
+We recommend you spend around a week researching the language and all the different components of the language before even beginning to modify and add the files below. Look at all the files in the project such as tokenizers, configuration files, grammar files, and different lookup functions to see what you need. This will save you a lot of time in the end. We highly suggest you stray away from ha ...
```
It seems like this paragraph is incomplete. Please finish the thought.

grhoten commented on inflection/resources/org/unicode/inflection/features/grammar.xml:
```diff
@@ -1624,6 +1624,106 @@
+            <grammeme name="singular"/>
+            <grammeme name="plural"/>
+        </category>
+        <category name="animacy">
+                <grammeme name="animate"/>
+                <grammeme name="inanimate"/>
+                <grammeme name="human"/>
+        </category>
```
Please use the same indentation level as the other categories. The indentation is 4 spaces.

grhoten commented on inflection/resources/org/unicode/inflection/tokenizer/config_ml.properties:
```diff
@@ -0,0 +1,7 @@
+#
+tokenizer.implementation.class=DefaultTokenizer
+tokenizer.nonDecompound.file=/org/unicode/inflection/tokenizer/ml/nondecompound.tok
+tokenizer.decompound=^(ശ്രീ)(.+?)(ഗുരു|സര്‍ക്കാര്‍)$|^(.+?)(ഗുരു|സര്‍ക്കാര്‍)$|^(.+?)(ഉണ്ട്|ആണ്|ഇല്ല)$|^(.+?)(ഒടൊപ്പം|ഉടൻ|ഓടെ|ഓട്|ഒപ്പം|തന്നെ|പോലും|പോലെ|ഉം|യ്)$|^(.+?)(കളുടെ|ങ്ങളുടെ|ത്തിന്റെ|ൻ്റെ|ന്റെ|യുടേ|യുടെ|യാൽ|യിൽ|ഇൽ|ല്|ൽ|ക്ക്|മാർ|ങ്ങൾ|കൾ|നെ|യെ)$
```
The ^ and $ should be removed. Those are automatically added by TokenExtractor. You'd be correct if those weren't automatically added.

grhoten commented on inflection/src/inflection/dialog/language/MlCommonConceptFactory.cpp:
```diff
@@ -0,0 +1,69 @@
+
+namespace inflection::dialog::language {
+
+MlCommonConceptFactory::MlCommonConceptFactory(const ::inflection::util::ULocale& language)
```
Is anything needed beyond the default constructor and destructor? This code looks a lot like the same behavior as EnCommonConceptFactory. If it's the default behavior, then please remove the redundant code.

grhoten commented on inflection/src/inflection/grammar/synthesis/MlGrammarSynthesizer_CaseLookupFunction.cpp:
```diff
@@ -0,0 +1,57 @@
+bool ends_with(const std::u16string& str, const std::u16string& suffix) {
+   if (suffix.size() >= str.size()) return false;
+   return std::equal(suffix.rbegin(), suffix.rend(), str.rbegin());
+}
```
Instead of writing this code, please consider using the ends_with on the u16string/u16string_view itself. Your `guessFallbackNounInflection` implementation has an example usage of using the string's ends_with.

grhoten commented on inflection/src/inflection/grammar/synthesis/MlGrammarSynthesizer_CaseLookupFunction.cpp:
```diff
@@ -0,0 +1,57 @@
+   std::u16string displayString;
+   ::inflection::util::StringViewUtils::lowercase(&displayString, displayValue.getDisplayString(), ::inflection::util::LocaleUtils::MALAYALAM());
+
+   static const std::vector<std::pair<std::u16string, std::u16string>> suffixToCase = {
```
There are several compiler warnings on Ubuntu about exit-time destructors. Please fix all of them. This is important to reduce the time to load and unload the library.

The best way to fix them is to move the all of them to data members of the relevant class. That way the memory being used can be unloaded when it's no longer in use.

Also if the value is constant, like the first value of the pair, it can be switched from u16string to a u16string_view. This requires less memory and it is constructed a little faster because less memory is allocated.

grhoten commented on inflection/src/inflection/grammar/synthesis/MlGrammarSynthesizer_MlDisplayFunction.cpp:
```diff
@@ -0,0 +1,381 @@
+    static constexpr auto TENSE_FUTURE = u"future";
+    static constexpr auto MOOD_INDICATIVE = u"indicative";
+    static constexpr auto MOOD_IMPERATIVE = u"imperative";
+    static constexpr auto MOOD_SUBJUNCTIVE = u"subjunctive";
```
Please reuse GrammemeConstants as much as possible.

grhoten commented on inflection/src/inflection/grammar/synthesis/MlGrammarSynthesizer_MlDisplayFunction.cpp:
```diff
@@ -0,0 +1,381 @@
+        (void)number;
+        (void)case_;
+        return u"";
+    }
```
This seems to be a no-op. Is this really needed, or are you planning on adding additional functionality?

grhoten commented on inflection/src/inflection/grammar/synthesis/MlGrammarSynthesizer_MlDisplayFunction.cpp:
```diff
@@ -0,0 +1,381 @@
+        {0x08 | 0x20, u"കളുടെ"},
+    };
+
+    static const std::map<int32_t, std::u16string> malayalamVerbSuffixMap = {
```
Please refrain from using magic numbers. Other implementations will use enums so that it's easier to read. It also makes it easier to maintain if a new value needs to be added. I believe Arabic, German and Italian do use enums to use in a key for a map.

It might be better to move these mappings to data members of the class instead too.

grhoten commented on inflection/src/inflection/grammar/synthesis/MlGrammarSynthesizer_MlDisplayFunction.cpp:
```diff
@@ -0,0 +1,381 @@
+        const dialog::DictionaryLookupInflector& dictionaryInflector) {
+        std::unique_ptr<inflection::tokenizer::Tokenizer> tokenizer(
+            inflection::tokenizer::TokenizerFactory::createTokenizer(util::LocaleUtils::MALAYALAM()));
+        if (!tokenizer) return std::nullopt;
```
please expand all if statements onto multiple lines. It's really hard to determine code coverage to ensure adequate testing when the if statement is a single line.  Using curly braces also make it easier to read, especially if the indentation is ever off.

grhoten commented on inflection/src/inflection/grammar/synthesis/MlGrammarSynthesizer_MlDisplayFunction.cpp:
```diff
@@ -0,0 +1,381 @@
+        int lastIndex = -1;
+        int64_t wordGrammemes = 0;
+        for (int i = 0; i < tokenChain->getWordCount(); ++i) {
+            std::u16string token = tokenChain->getTokenValue(i);
```
Normally, this is simplified to the following. Please use this instead. It's more efficient. You can check if the token is significant by checking the result of `token.isSignificant()`.
```
    for (const auto& token : *tokenChain) {
        const auto& tokenVal = token.getValue();
```

`ArGrammarSynthesizer_ArDisplayFunction::inflectString()` has a more complete example of how to make this code shorter.

grhoten commented on inflection/src/inflection/grammar/synthesis/MlGrammarSynthesizer_MlDisplayFunction.cpp:
```diff
@@ -0,0 +1,381 @@
+
+        std::u16string result;
+        for (size_t i = 0; i < tokens.size(); ++i) {
+            if (i > 0) result += u" ";
```
It's not always a space. I recommend checking `ArGrammarSynthesizer_ArDisplayFunction::inflectString()` for how to handle such situations.

grhoten commented on inflection/src/inflection/grammar/synthesis/MlGrammarSynthesizer_MlDisplayFunction.cpp:
```diff
@@ -0,0 +1,381 @@
+                    // Rebuild full phrase for noun fallback
+                    std::u16string fullPhrase;
+                    for (size_t i = 0; i < tokens.size(); ++i) {
+                        if (i > 0) fullPhrase += u" ";
```
It's not always a space. Please check `ArGrammarSynthesizer_ArDisplayFunction::inflectString()` for an example of handling all of the irrelevant stuff.

grhoten commented on inflection/src/inflection/grammar/synthesis/MlGrammarSynthesizer_MlDisplayFunction.cpp:
```diff
@@ -0,0 +1,381 @@
+#include <vector>
+
+namespace inflection::grammar::synthesis {
+    using dialog::SemanticFeature;
```
All of the other code doesn't inflect this far. Please use similar indentation as the other classes.

grhoten commented on inflection/src/inflection/grammar/synthesis/MlGrammarSynthesizer_MlDisplayFunction.cpp:
```diff
@@ -0,0 +1,381 @@
+    using dialog::SemanticFeatureModel_DisplayData;
+    using dialog::DisplayValue;
+
+    static const icu4cxx::UnicodeSet malayalamInflectableChars(u"[\\u0D00-\\u0D7F]");
```
Can you make this into a more readable and reusable set? For example, StringFilterUtil has several script based sets. Perhaps you can define `MALAYALAM_SCRIPT` as `[:Malayalam:]` in that class.

grhoten commented on inflection/src/inflection/grammar/synthesis/MlGrammarSynthesizer_MlDisplayFunction.cpp:
```diff
@@ -0,0 +1,381 @@
+    using dialog::DisplayValue;
+
+    static const icu4cxx::UnicodeSet malayalamInflectableChars(u"[\\u0D00-\\u0D7F]");
+    static const icu4cxx::UnicodeSet nonMalayalamChars(u"[\\p{Latin}\\p{Nd}\\p{Punct}]");
```
I suspect that this is redundant with the Malayalam script above. What's the thought process for using this set?

grhoten commented on inflection/src/inflection/grammar/synthesis/MlGrammarSynthesizer_MlDisplayFunction.cpp:
```diff
@@ -0,0 +1,381 @@
+        return result;
+    }
+
+    MlGrammarSynthesizer_MlDisplayFunction::MlGrammarSynthesizer_MlDisplayFunction(
```
Can you move this to the beginning like the other implementations? It makes it easier to review and switch between implementations.

grhoten commented on inflection/src/inflection/grammar/synthesis/MlGrammarSynthesizer_MlDisplayFunction.cpp:
```diff
@@ -0,0 +1,381 @@
+          tenseFeature(*npc(model.getFeature(u"tense"))),
+          moodFeature(*npc(model.getFeature(u"mood"))),
+          pronounTypeFeature(*npc(model.getFeature(u"pronounType"))),
+          determinationFeature(*npc(model.getFeature(u"determination"))),
```
Please consider switching all of the hard coded strings to GrammemeConstants. I like reuse.

grhoten commented on inflection/src/inflection/grammar/synthesis/MlGrammarSynthesizer_MlDisplayFunction.cpp:
```diff
@@ -0,0 +1,381 @@
+                  {CLUSIVITY_INCLUSIVE, CLUSIVITY_EXCLUSIVE},
+                  {PERSON_FIRST, PERSON_SECOND, PERSON_THIRD},
+                  {TENSE_PAST, TENSE_PRESENT, TENSE_FUTURE},
+                  {MOOD_INDICATIVE, MOOD_IMPERATIVE, MOOD_SUBJUNCTIVE}
```
Looks good. Do adjectives inflect?  I thought you said no previously. If so, then nothing more needs to change for these arguments.

grhoten commented on inflection/src/inflection/grammar/synthesis/MlGrammarSynthesizer_NumberLookupFunction.cpp:
```diff
@@ -0,0 +1,71 @@
+        npc(npc(tokenizer.get())->createTokenChain(word)));
+
+    for (const auto& token : *tokenChain) {
+        if (dynamic_cast<const ::inflection::tokenizer::Token_Word*>(&token) != nullptr) {
```
Or you could check `token.isSignificant()`.  It's a little clearer when using that. You probably copied this code from English or a similar implementation. It's probably better to use isSignificant instead.

grhoten commented on inflection/src/inflection/tokenizer/locale/ml/MlTokenizer.hpp:
```diff
@@ -0,0 +1,17 @@
+#include <inflection/tokenizer/TokenExtractor.hpp>   
+#include <map>
+
+class inflection::tokenizer::locale::ml::MlTokenizer final
```
Please remove the changes to tokenizer/locale/ml. They seem to be unused and not needed.

grhoten commented on inflection/test/resources/inflection/dialog/inflection/ml.xml:
```diff
@@ -0,0 +1,73 @@
+
+<!-- Malayalam Adjective Inflection Tests -->
+<test><source gender="feminine">പുതിയ</source><result>പുതിയ</result></test>
+<test><source gender="masculine">പുതിയ</source><result>പുതിയ</result></test>
```
Are these results changing?  Am I missing something? What's being tested here? If it's not supposed to change on purpose, please add a comment about that to this section of the tests.

Also please indent this file like the other xml files. It makes it easier to navigate between them.

grhoten reviewed (changes requested):
Overall, the changes look about right. Some code cleanup is still needed to improve maintainability by others, and to handle edge cases.

Here are some additional tasks to work on before merging these changes.

- [ ] Please switch the dictionary_ml.lst and inflectional_ml.xml back to git lfs. It should be working again.
- [ ] Please switch the title of this pull request to a more informative summary, like "Inflection 85: Add support for Malayalam". You can also reference the issue number with "Fixes" so that the issue will be automatically closed when this is merged.
- [ ] Please update the summary to summarize what functionality is being implemented. The summary of the current state from your Google doc would be a good to put here.
- [ ] Please git squash these changes. The intermediate states aren't needed.

What hasn't been contributed are the RBNF rules for Malayalam. That is used by the quantify method of CommonConceptFactory. If you would like to contribute or work on that, that can be a separate submission to CLDR. I can go over those details, if you want to know more. This isn't a prerequisite for these changes, but it does help with the overall completeness of the implementation.

Thanks for doing this!

deonajulary06 committed:
Changed files based on comments

deonajulary06 committed:
Inflection 85: Add language support and grammar synthesis for Malayalam

Inflection 85: Add language support and grammar synthesis for Malayalam

Add Malayalam dictionary support

Add Malayalam language support in LocaleUtils.hpp

Add Malayalam locale support to LocaleUtils

Add Malayalam language to the tests

Add Malayalam locale group ml_IN

ADD: Malayalam tokenizer configuration file

Inflection-85: Add Git LFS config for Malayalam dictionary and XML files

Add Malayalam inflection and pronoun tests

Updated copyright line

Updated copyright message

Updated copright message

Updated language grammar to include Malayalam

Added pronouns for Malayalam

Add ll GrammarSynthesizer files

Add Malayalam grammar synthesizer

Add Malayalam-specific CommonConceptFactory with lists and quantities

Update document on how to add a new language, fixed errors

Updated grammar.xml for Malayalam

Update pronoun_ml.csv

Updated all grammar synthesizer component for Malayalam

Update Common Concept Factory files

Updated tests for Malayalam

Fix Malayalam grammar synthesis and remove count lookup function

Updated Grammeme Constants files to include sociative case

Temporary fix for GitHub

Modified files to fix more test errors

Modified files to fix more test errors

Update files to fix errors

Same file as before but with corrected indentations

Update test files to reflect tokenization

Added tokenizer files

Added feedback on how to add a new language

Modified files

New changes to fix errors

Updated Grammar Synthesizer file

Fixed errors

Made changes to fix errors

Fix Common Concept Factory

Reduce memory leaks

deonajulary06 committed:
Quick fix

deonajulary06 committed:
Merge branch 'main' into update-new-language-doc

deonajulary06 committed:
Restore Malayalam dictionary + inflectional data after LFS migrate

grhoten commented:
The remainder of the review will be moved to the new pull request 175.

grhoten commented:
This has been resubmitted as #176
